packages/@aws-cdk/core/lib/private/physical-name-generator.ts

```diff
@@ -65,3 +67,26 @@ class SuffixNamePart extends NamePart {
     return this.bareStr.slice(startIndex, strLen);
   }
 }
+
+const GENERATE_IF_NEEDED_SYMBOL = Symbol.for('@aws-cdk/core.<private>.GenerateIfNeeded');
+
+export class GeneratedWhenNeededMarker implements IResolvable {
```

Comment:
Feels like we can perhaps generalize the idea of "marker tokens" but we'll do that next time this need arises.

Reply:
add some commentary on how this works